### PR TITLE
[backend]: add 'creator' and 'from' query params to notes endpoint

### PR DESCRIPTION
**Pull request by Haypierre**

### CODE REVIEW

File: fc-community-backend/apps/api/src/factchain-core/web3.ts [hunk 1]
```diff
@@ -74,37 +69,35 @@ export class FactChainBackend implements NoteReader, NoteWriter {
...
     );
-    let notePromises: Promise<Note>[] = [];
+    let notes: Note[] = [];
+
     for (const period of block_periods) {
       const events = await this.getEvents("NoteCreated", period[0], period[1]);
+      const allCreatedNotes = await Promise.all(
+        events.map(async (event) => {
+          return await this.getNote(event.args[0], event.args[1]);
+        }),
```
Comment (YBadiss) [inline fc-community-backend/apps/api/src/factchain-core/web3.ts:93]: I see what you mean about the issue of selecting all, and only then filtering.

Maybe we can keep things simple and efficient for now by expecting a `predicate: (postUrl, creatorAddress) => boolean` ?
This way you can still do

```javascript
const relatedEvents = events.filter((e) => predicate(e.args[0], e.args[1]));
if (relatedEvents) {
  notePromises = notePromises.concat(
    relatedEvents.map(async (event) => {
      return await this.getNote(event.args[0], event.args[1]);
    }),
  );
}
...
```

File: fc-community-backend/apps/api/src/factchain-core/web3.ts [hunk 2]
```diff
@@ -74,37 +69,35 @@ export class FactChainBackend implements NoteReader, NoteWriter {
     };
   };
 
-  getNotes = async (postUrl: string): Promise<Array<Note>> => {
+  getNotes = async (
+    predicate: (note: Note) => boolean,
```
Comment (YBadiss) [inline fc-community-backend/apps/api/src/factchain-core/web3.ts:73]: can you rename `predicate` to something more meaningful? Even something basic like `filterFn`.

File: fc-community-backend/apps/api/src/app.service.ts [hunk 3]
```diff
@@ -14,10 +16,30 @@ export class AppService {
...
+      (note: Note) => note.postUrl === postUrl,
+      from,
+    );
+    return notes;
+  }
+
+  async getAllNotesFrom(from: number): Promise<Array<Note>> {
+    const fc = new FactChainBackend(config);
+    const ns = new NoteService(fc, fc);
+    const notes = await ns.getNotes((note: Note) => note === note, from);
```
Comment (YBadiss) [inline fc-community-backend/apps/api/src/app.service.ts:32]: if you change the type of predicate to be `<the function type> | undefined` then you don't need to create a predicate here that is always true.

If you don't like that, let's do something like `(note: Note) => true` instead.

**Review by YBadiss — CHANGES_REQUESTED** [review]

File: fc-community-backend/apps/api/src/app.controller.ts [hunk 4]
```diff
@@ -20,9 +24,32 @@ export class AppController {
...
-  async getNotes(@Query("postUrl") postUrl): Promise<NotesResponse> {
-    console.log(`Get notes for postUrl=${postUrl}`);
-    const notes = await this.appService.getNotes(postUrl);
+  async getNotes(
+    @Query("postUrl") postUrl: string,
+    @Query("from") from: number,
+    @Query("creator") address: string,
+  ): Promise<NotesResponse> {
+    let notes = [];
+    let parsedFrom = 0;
```
[the remark this answers is not in the document]
Comment (Haypierre) [reply fc-community-backend/apps/api/src/app.controller.ts:33]: oops

File: fc-community-backend/apps/api/src/factchain-core/web3.ts (hunk 2 above)
Comment (Haypierre) [reply fc-community-backend/apps/api/src/factchain-core/web3.ts:73]: Screenshot 2024-01-04 at 16 00 20


predicate is the conventional naming for higher order function given to filter.

File: fc-community-backend/apps/api/src/app.service.ts (hunk 3 above)
Comment (Haypierre) [reply fc-community-backend/apps/api/src/app.service.ts:32]: `If you don't like that, let's do something like (note: Note) => true instead.`

matter of style ... I was first going for `(note: Note) => true` then got triggered by the "unused note parameter"

**Pull request merged**